README.md
# jFloatWavIO
[![Latest Release](https://gitlab.com/f-matano44/jfloatwavio/-/badges/release.svg)](https://gitlab.com/f-matano44/jfloatwavio/-/releases) [![Build status](https://gitlab.com/f-matano44/jfloatwavio/badges/main/pipeline.svg)](https://gitlab.com/f-matano44/jfloatwavio/-/jobs) [![](https://jitpack.io/v/com.gitlab.f-matano44/jfloatwavio.svg)](https://jitpack.io/#com.gitlab.f-matano44/jfloatwavio)<br>
Java package to read/write wav file as float or double array. <br>
It is not compatible between version 0 and version 1.

* Repository & JitPack source: https://gitlab.com/f-matano44/jfloatwavio
* Mirror repository: https://github.com/f-matano44/jfloatwavio


## Development environment
* [Gradle](https://gradle.org/) + [Java 17 (LTS)](https://adoptium.net/temurin/releases/?version=17)
* [VSCode](https://code.visualstudio.com/) + [Checkstyle for Java](https://marketplace.visualstudio.com/items?itemName=shengchen.vscode-checkstyle)

Part of this Library is made that using ChatGPT (GPT-4).


## Readable/writeable format
| \ |value|
|---|-----|
|Encoding|PCM_SIGNED|
|Channels|1 (mono) or 2 (stereo)|
|Sample rate|any|
|Bit depth|8, 16, 24, 32|
|Endian|any|


## Usage
### Add dependency 
#### module-info.java
```java
module module_name {
    ...
    requires jp.f_matano44.jfloatwavio;
}
```
#### gradle.build.kts
Please replace {version} with the desired version name (greater than 1.3.1).
```kotlin
repositories {
    ...
    maven { url = uri("https://jitpack.io") }
}

dependencies {
    ...
    implementation("com.gitlab.f-matano44:jfloatwavio:{version}")
}
```

### Functions
```Java
import jp.f_matano44.jfloatwavio.*;
import javax.sound.sampled.AudioFormat;

// If you use static functions...
double[][] signals = sGetSignal("path/to/inputFile.wav");
double[] left_or_mono = signal[0];
double[] right = signal[1];
AudioFormat format = sGetFormat("path/to/inputFile.wav");
int nbits = format.getSampleSizeInBits();
int fs = (int) format.getSampleRate();
sOutputData("path/to/outputFile.wav", nbits, fs, mono);
sOutputData("path/to/outputFile.wav", nbits, fs, left, right);

// else if you use WavIO object...
try {
    // input
    WavIO inWio = new WavIO("path/to/inputFile.wav");
    double[] left_or_mono = inWio.getSignal()[0];
    double[] right = inWio.getSignal()[1];
    AudioFormat format = inWio.getFormat();

    // output
    WavIO outWio = new WavIO(format, mono);
    WavIO outWio = new WavIO(foramt, left, right);
    outWio.outputData("path/to/outputFile.wav");
} catch (Exception e) {
    // Error handling
}
```


## For developer
### Build & Publish to local repository (mavenLocal)
```SH
jfloatwavio/$ ./gradlew publish
```

### Add dependency (gradle.build.kts)
Please replace {version} with the name of the built version.
```kotlin
repositories {
    ...
    mavenLocal()
}

dependencies {
    ...
    implementation("jp.f-matano44:jfloatwavio:{version}")
}
```


# COPYRIGHT
Copyright 2023 Fumiyoshi MATANO<br>


CHANGELOG
1.4.0
* Compatible with Module System

1.3.2
* bugfix

1.3.1
* mavenLocal -> JitPack.io

1.3.0
* ant -> gradle

1.2.1
* Bug fix

1.2.0
* Changed access modifier of Converter (between byte and double) method from private to public
* Fixed a bug preventing the proper reading of big-endian files
* Corrected an issue causing a reduction in read accuracy under specific conditions
* Updated Javadoc to include more detailed information

1.1.2
* modify filename

1.1.1
* Update documents

1.1.0
* Added examples for static method to Test.java
* Refining the exceptions being thrown
* Bug fix

1.0.0
* Delete "DrawSignal" method
* Add static method

0.0.2
* Delete debug code

0.0.1
* First Release


test/Test.java
import java.io.IOException;
import javax.sound.sampled.AudioFormat;
import jp.f_matano44.jfloatwavio.WavIO;

/** test program. */
public class Test {
    /** main method of test program. */
    public static void main(String[] args) {
        WavIO input1 = null;
        WavIO output = null;

        /* 
         * input wav file 
         * usage: WavIO WavIO_Obj = new WavIO(String FILENAME);
         */
        try {
            input1 = new WavIO("zundamon.wav");
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }


        /* get format */
        final AudioFormat input1Format = input1.getFormat();
        final float fs = input1Format.getSampleRate();
        final int nbits = input1Format.getSampleSizeInBits();


        /* 
         * get {float, double} array
         *      signal[0]: monoral or left
         *      signal[1]: right
         * 
         * usage
         *  double[][]  signal = wavIO_Obj.getSignal();
         *  float[][]   signal = wavIO_Obj.getFloatSignal();
         * 
         * usage (static method)
         *  double[][]  signal = WavIO.sGetSignal();
         *  float[][]   signal = WavIO.sGetFloatSignal();
         */
        final double[] x = input1.getSignal()[0];
        final double[] y = WavIO.sGetSignal("metan.wav")[0];


        // set format
        final int outputChannels = 2;
        final float outputFs = fs;
        final int outputNbits = nbits;
        final boolean signed = true; // This library is allowed PCM-SIGN only.
        final boolean bigEndian = false;
        final AudioFormat outputFormat = new AudioFormat(
            (float) outputFs, outputNbits, outputChannels, signed, bigEndian
        );


        /* 
         * WavIO object can be generated from signal. 
         * If not supported format, throws Exception.
         * This call requires that the arrays of argument are same length.
         * 
         * usage
         *  WavIO WavIO_Obj = new WavIO(AudioFormat format, double[]... signal);
         */
        try {
            output = new WavIO(outputFormat, x, y);
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }


        /* 
         * print format info
         * usage: WavIO_Obj.printAudioFormat();
         */
        output.printAudioFormat();


        /* 
         * output wav file 
         *
         * usage
         *  wavIO_Obj.outputData(String FILENAME);
         * 
         * usage (static method)
         *  WavIO.outputData(String FILENAME, int nbits, double fs, double[]... signal);
         *      signal[0]: left or mono
         *      signal[1]: right
        */
        try {
            output.outputData("output01.wav");
            WavIO.sOutputData("output02.wav", outputNbits, outputFs, x, y);
        } catch (IOException e) {
            e.printStackTrace();
            System.out.println("Failed of writing file.");
        }
    }
}


src/test/Test.java
import java.io.IOException;
import javax.sound.sampled.AudioFormat;
import jp.f_matano44.jfloatwavio.*;

public class Test
{
    public static void main(String[] args)
    {
        AudioFormat inputFormat, outputFormat;
        WavIO input1=null, input2=null, output=null;
        double[] x, y;


        /* 
         * input wav file 
         * 
         * usage
         *  WavIO WavIO_Obj = new WavIO(String FILENAME);
         */
        try{
            input1 = new WavIO("zundamon.wav");
            input2 = new WavIO("metan.wav");
        }catch(Exception e){
            e.printStackTrace();
            System.exit(1);
        }


        /* 
         * get double array 
         * 
         * usage
         *  double[] signal = wavIO_Obj.getX();
         *  float[]  signal = wavIO_Obj.getXf();
         */
        x = input1.getX()[0];
        y = input2.getX()[0];


        // set format
        inputFormat = input1.getFormat();
        outputFormat = new AudioFormat(
            AudioFormat.Encoding.PCM_SIGNED, inputFormat.getSampleRate(),
            inputFormat.getSampleSizeInBits(), 2, inputFormat.getFrameSize()*2,
            inputFormat.getFrameRate(), false
        );


        /* 
         * WavIO object can be generated from signal. 
         * If not supported format, throws Exception.
         * This call requires that the arrays of argument are same length.
         * 
         * usage
         *  WavIO WavIO_Obj = new WavIO(AudioFormat format, double[]... signal);
         */
        try{
            output = new WavIO(outputFormat, x, y);
        }catch(Exception e){
            e.printStackTrace();
            System.exit(1);
        }


        /* 
         * draw waveform 
         * 
         * usage
         *  new DrawSignal(String WindowTitle, double[]... signal);
         */
        // draw original waveform
        new DrawSignal("inputSignal 1", x);
        new DrawSignal("inputSignal 2", y);
        // draw connected waveform
        new DrawSignal("outputSignal", output.getX());


        /* 
         * print format info
         * 
         * usage
         *  WavIO_Obj.printAudioFormat();
         */
        output.printAudioFormat();

        /* 
         * output wav file 
         *
         * usage
         *  wavIO_Obj.outputData(String FILENAME);
        */
        try{
            input1.outputData("test1.wav");
            output.outputData("helloworld.wav");
        }catch(IOException e){
            e.printStackTrace();
            System.out.println("Failed of writing file.");
        }
    }
}


lib/src/test/java/jp/f_matano44/jfloatwavio/ConverterTestForRandom.java
package jp.f_matano44.jfloatwavio;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;

import java.util.Random;
import org.junit.jupiter.api.Test;

/** Unit test for random case. */
class ConverterTestForRandom {
    private static void randomCheck(int nBits, boolean isBigEndian) {
        final byte[] expected = new byte[nBits * 100000];
        new Random(System.currentTimeMillis()).nextBytes(expected);
        final double[] doubleArray = Converter.byte2double(expected, nBits, isBigEndian);
        final byte[] actual = Converter.double2byte(doubleArray, nBits, isBigEndian);

        assertArrayEquals(expected, actual);
    }


    // ----------------------------------------
    // nBits = 8
    // ----------------------------------------
    @Test void randomCase_08bit_BigEndian() {
        final int nBits = 8;
        final boolean isBigEndian = true;
        randomCheck(nBits, isBigEndian);
    }

    @Test void randomCase_08bit_LittleEndian() {
        final int nBits = 8;
        final boolean isBigEndian = false;
        randomCheck(nBits, isBigEndian);
    }


    // ----------------------------------------
    // nBits = 16
    // ----------------------------------------
    @Test void randomCase_16bit_BigEndian() {
        final int nBits = 16;
        final boolean isBigEndian = true;
        randomCheck(nBits, isBigEndian);
    }

    @Test void randomCase_16bit_LittleEndian() {
        final int nBits = 16;
        final boolean isBigEndian = false;
        randomCheck(nBits, isBigEndian);
    }


    // ----------------------------------------
    // nBits = 24
    // ----------------------------------------
    @Test void randomCase_24bit_BigEndian() {
        final int nBits = 24;
        final boolean isBigEndian = true;
        randomCheck(nBits, isBigEndian);
    }

    @Test void randomCase_24bit_LittleEndian() {
        final int nBits = 24;
        final boolean isBigEndian = false;
        randomCheck(nBits, isBigEndian);
    }


    // ----------------------------------------
    // nBits = 32
    // ----------------------------------------
    @Test void randomCase_32bit_BigEndian() {
        final int nBits = 32;
        final boolean isBigEndian = true;
        randomCheck(nBits, isBigEndian);
    }

    @Test void randomCase_32bit_LittleEndian() {
        final int nBits = 32;
        final boolean isBigEndian = false;
        randomCheck(nBits, isBigEndian);
    }
}


src/jp/f_matano44/jfloatwavio/WavInput.java
package jp.f_matano44.jfloatwavio;

import java.nio.ByteBuffer;
import javax.sound.sampled.AudioFormat;

abstract class WavInput
{
    // variables
    protected AudioFormat format;
    protected int fs;       // sample rate
    protected double[][] x; // signal data
    // constants
    protected static final int intIs4Bytes = 4;


    // methods
    protected double[] byte2double(
        final byte[] byteArray, final int nBits, final boolean isBigEndian)
    {
        final int
            nBytes = nBits / 8,
            sampleNum = byteArray.length / nBytes;
        double[] doubleArray = new double[sampleNum];
        byte[] temp = new byte[intIs4Bytes];

        for(int i=0; i<byteArray.length; i++)
        {
            final int
                tPos = i % nBytes,
                dPos = i / nBytes;

            temp[tPos] = byteArray[i];
            if(tPos % nBytes == nBytes - 1)
            {
                // preProcess
                temp = getBigEndian(temp, isBigEndian);
                temp = fillArray(temp, nBytes);
                // byte to int
                doubleArray[dPos] = (double)ByteBuffer.wrap(temp).getInt();
                // int to double
                doubleArray[dPos] = (double)doubleArray[dPos] / Math.pow(2, nBits);
            }
        }

        return doubleArray;
    }


    protected byte[] getBigEndian(byte[] array, final boolean isBigEndian)
    {
        if(!isBigEndian)
        {
            final int tail = array.length - 1;
            for(int i=0; i<(array.length/2); i++)
            {
                byte temp;
                temp = array[i];
                array[i] = array[tail-i];
                array[tail-i] = temp;
            }
        }
        return array;
    }


    protected byte[] fillArray(byte[] array, int nBytes)
    {
        final int 
            fillDigit = intIs4Bytes-nBytes,
            fillNum = (array[fillDigit] >> 7) & 1;
        for(int i=0; i<fillDigit; i++)
        {
            if(fillNum == 0)
                array[i] = 0;
            else
                array[i] = -1;
        }
        return array;
    }


    public void printAudioFormat()
    {
        System.out.println("Encoding: " + this.format.getEncoding());
        System.out.println("isBigEndian: " + this.format.isBigEndian());        
        System.out.println("Channels:" + this.format.getChannels());
        System.out.println("");
        System.out.println("SampleRate: " + this.format.getSampleRate());
        System.out.println("BitDepth: " + this.format.getSampleSizeInBits());
        System.out.println("");
        System.out.println("FrameRate: " + this.format.getFrameRate());
        System.out.println("FrameSize: " + this.format.getFrameSize());
    }
}


src/jp/f_matano44/jfloatwavio/MonoInput.java
package jp.f_matano44.jfloatwavio;

import java.io.File;
import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioSystem;


public class MonoInput extends WavInput
{
    public int getFs(){ return super.fs; }          // getter of sampling rate
    public double[] getX(){ return super.x[0]; }    // getter of signal date (double)
    public float[] getXf(){                         // getter of signal date (float)
        float[] xf = new float[super.x.length];
        for(int i=0; i<super.x.length; i++)
            xf[i] = (float)super.x[0][i];
        return xf;
    }


    public MonoInput(String FILENAME)
    {
        final File f = new File(FILENAME);
        byte[] sByteArray = null;
        AudioFormat sFormat = null;
        int channels, xLength;

        // import wav file data
        try (var s = AudioSystem.getAudioInputStream(f))
        {
            sByteArray = s.readAllBytes();
            sFormat = s.getFormat();

            // reject don't allowed files
            if(sFormat.getChannels() != 1)
                throw new Exception("MonoInput is allowed monoral only..");
        }
        catch (Exception e)
        {
            e.printStackTrace();
            System.exit(1);
        }

        channels = sFormat.getChannels();
        xLength = sByteArray.length/(8 * sFormat.getSampleSizeInBits());
        super.x = new double[channels][xLength];
        // get signal data
        super.x[0] = byte2double(sByteArray, sFormat.getSampleSizeInBits(), sFormat.isBigEndian());
        // get sampling rate
        super.fs = (int)sFormat.getSampleRate();
        // get format info
        super.format = sFormat;
    }
}


src/jp/f_matano44/jfloatwavio/DrawSignal.java
package jp.f_matano44.jfloatwavio;

import java.awt.*;
import javax.swing.*;

public class DrawSignal extends JFrame
{
    public DrawSignal(String windowTitle, double[]... signal)
    {
        final int 
            defaultWidth = 600,
            defaultHeight = 40 + 160 * signal.length;
        // window setting
        setTitle(windowTitle);
        setBackground(Color.WHITE);
        setSize(defaultWidth, defaultHeight);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        // add graph panel
        getContentPane().add(this.new PanelClass(signal));
        // display panel
        setVisible(true);
    }


    private class PanelClass extends JPanel
    {
        private double[][] x;
        public PanelClass(final double[][] signal){ this.x = signal; }

        @Override
        public void paint(final Graphics g)
        {
            // canvas size
            final int frameWidth = 10;
            final Dimension d = getSize();
            final int 
                windowWidth = (int)d.getWidth(),
                windowHeight = (int)d.getHeight(),
                minWindowSize = frameWidth + frameWidth*x.length,
                canvasWidth = (int)d.getWidth() - frameWidth * 2,
                canvasHeight = (((int)d.getHeight() - frameWidth) / x.length) - frameWidth;

            // clear frame
            g.clearRect(0, 0, windowWidth, windowHeight);

            // if window is too narrow, display nothing.
            if(windowWidth < minWindowSize || windowHeight < minWindowSize)
                return;

            // draw panel
            for(int i=0; i<x.length; i++)
            {
                final int xAxis = 0, yAxis=1, startPoint=0, endPoint=1;
                final int
                    xOffset = frameWidth,
                    yOffset = frameWidth + canvasHeight/2 + ((frameWidth+canvasHeight)*i);
                int[][] pos = new int[2][2];
                    pos[xAxis][endPoint] = xOffset;
                    pos[yAxis][endPoint] = yOffset;

                // draw background
                g.setColor(Color.LIGHT_GRAY);
                g.fillRect(
                    frameWidth, frameWidth + (canvasHeight + frameWidth) * i,
                    canvasWidth, canvasHeight
                );

                // draw signal
                g.setColor(Color.BLACK);
                for(int j=0; j<x[i].length; j++)
                {
                    final double AMP_MAX = (canvasHeight - (frameWidth*2)) / 2;

                    pos[xAxis][startPoint] = pos[xAxis][endPoint];
                    pos[yAxis][startPoint] = pos[yAxis][endPoint];
                    pos[xAxis][endPoint] =
                        xOffset + (int)(canvasWidth * ((double)j / (x[i].length-1)));
                    pos[yAxis][endPoint] =
                        yOffset + (int)((-1) * x[i][j] * AMP_MAX);

                    if(0 < AMP_MAX)
                    {
                        g.drawLine(
                            pos[xAxis][startPoint], pos[yAxis][startPoint],
                            pos[xAxis][endPoint], pos[yAxis][endPoint]
                        );
                    }
                }
            }
        }
    }
}


lib/src/main/java/jp/f_matano44/jfloatwavio/WavIO.java
package jp.f_matano44.jfloatwavio;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import javax.sound.sampled.AudioFileFormat;
import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.UnsupportedAudioFileException;


/** Java library for wav file as float. */
public class WavIO {
    public static void main(String[] args) {
        System.out.println("jFloatWavIO @ Mozilla Public License 2.0");
        System.out.println("Copyright 2023 Fumiyoshi MATANO");
    }


    /**
     * Retrieves the wave signal from the provided file and returns it as a 2D array of doubles.
     * This is a static method that can be used without instantiating the class.
     *
     * @param filename 
     *      The name of the file from which to read the wave signal.
     * @return
     *      A 2D array of doubles representing the wave signal.
     *      <ul>
     *          <li> success: double[][]
     *              <ul>
     *                  <li> signal[0]: left or mono </li>
     *                  <li> signal[1]: right </li>
     *              </ul>
     *          </li>
     *          <li> failure: null </li>
     *      </ul>
     */
    public static double[][] sGetSignal(final String filename) {
        try {
            return new WavIO(filename).getSignal();
        } catch (Exception e) {
            return null;
        }
    }


    /**
     * Retrieves the AudioFormat from the specified file.
     * This is a static method that can be used without instantiating the class.
     *
     * @param filename
     *      The name of the file from which to retrieve the AudioFormat.
     * @return
     *      The AudioFormat extracted from the file.
     *      <ul>
     *          <li> success: AudioFormat </li>
     *          <li> failure: null </li>
     *      </ul>
     */
    public static AudioFormat sGetFormat(final String filename) {
        try {
            return new WavIO(filename).getFormat();
        } catch (Exception e) {
            return null;
        }
    }


    /**
     * Writes the provided signal data to a .wav file.
     * This is a static method that can be used without instantiating the class.
     * Encoding: PCM_SIGNED
     * Endian: Little Endian
     *
     * @param filename The name of the output .wav file.
     * @param nbits The bit depth for the audio data.
     * @param fs The sampling rate of the audio data.
     * @param mono The audio signal data to be written.
     * 
     * @return status
     *      <ul>
     *          <li>  0: success </li>
     *          <li> -1: failure </li>
     *      </ul>
     */
    public static int sOutputData(
        final String filename, final int nbits, final double fs,
        final double[] mono
    ) {
        return sOutputDataBody(filename, nbits, fs, mono);
    }


    /**
     * Writes the provided signal data to a .wav file.
     * This is a static method that can be used without instantiating the class.
     * Encoding: PCM_SIGNED
     * Endian: Little Endian
     *
     * @param filename The name of the output .wav file.
     * @param nbits The bit depth for the audio data.
     * @param fs The sampling rate of the audio data.
     * @param left The audio signal data to be written.
     * @param right The audio signal data to be written.
     * 
     * @return status
     *      <ul>
     *          <li>  0: success </li>
     *          <li> -1: failure </li>
     *      </ul>
     */
    public static int sOutputData(
        final String filename, final int nbits, final double fs,
        final double[] left, final double[] right
    ) {
        return sOutputDataBody(filename, nbits, fs, left, right);
    }


    private static int sOutputDataBody(
        final String filename, final int nbits, final double fs,
        final double[]... signal
    ) {
        try {
            final int channels = signal.length;
            final AudioFormat outputFormat = new AudioFormat(
                    (float) fs, nbits, channels, true, false
                );

            if (isFormatOK(outputFormat)) {
                new WavIO(outputFormat, signal).outputData(filename);
                return 0;
            } else {
                throw new UnsupportedAudioFileException();
            }
        } catch (Exception e) {
            return -1;
        }
    }


    // member variable
    private final AudioFormat format;
    private final double[][] signal;


    /**
     * getter of audio format.
     *
     * @return AudioFormat (Immutable)
     */
    public AudioFormat getFormat() {
        return this.format;
    }


    /**
     * getter of signal data.
     *
     * @return signal data as double[][]
     *      <ul>
     *          <li> signal[0]: left or mono </li>
     *          <li> signal[1]: right </li>
     *      </ul>
     */
    public double[][] getSignal() {
        final double[][] x = new double[this.signal.length][];
        for (int i = 0; i < this.signal.length; i++) {
            x[i] = this.signal[i].clone();
        }
        return x;
    }


    /**
     * getter of signal data.
     *
     * @return signal data as float[][]
     *      <ul>
     *          <li> signal[0]: left or mono </li>
     *          <li> signal[1]: right </li>
     *      </ul>
     */
    public float[][] getFloatSignal() {
        final float[][] xf = new float[this.signal.length][];
        for (int i = 0; i < this.signal.length; i++) {
            xf[i] = new float[this.signal[i].length];
            for (int j = 0; j < this.signal[i].length; j++) {
                xf[i][j] = (float) this.signal[i][j];
            }
        }
        return xf;
    }


    /**
     * Constructs a new WavIO object from the provided signal array,
     * using the provided AudioFormat.
     *
     * @param f AudioFormat object specifying the data format of the audio data.
     * @param signal Array containing the audio signal data.
     *      <ul>
     *          <li> signal[0]: left or mono </li>
     *          <li> signal[1]: right </li>
     *      </ul>
     *
     * @throws UnsupportedAudioFileException
     *      if the provided AudioFormat is not supported.
     * @throws IllegalArgumentException
     *      if AudioFormat and Signal channel counts don't match.
     */
    public WavIO(final AudioFormat f, final double[]... signal)
        throws UnsupportedAudioFileException, IllegalArgumentException {
        // ファイルフォーマットが異常な場合に例外を投げる
        if (!isFormatOK(f)) {
            throw new UnsupportedAudioFileException(
                "This format is unsupported."
            );
        } else if (signal.length != f.getChannels()) {
            throw new IllegalArgumentException(
                "AudioFormat and Signal channel counts don't match."
            );
        }

        // メンバ変数にコピー
        this.format = f;
        this.signal = adjustLength(signal);
    }


    /**
     * Constructs a new WavIO object from the provided file name.
     *
     * @param filename the name of the file to be processed
     * @throws IOException if an error occurs during file reading
     * @throws UnsupportedAudioFileException if Provided file is not supported
     */
    public WavIO(final String filename) 
        throws UnsupportedAudioFileException, IOException {
        // wav の読み込み
        final File f = new File(filename);
        final var ais = AudioSystem.getAudioInputStream(f);

        // ファイルフォーマットが異常な場合に例外を投げる
        this.format = ais.getFormat();
        if (!isFormatOK(this.format)) {
            throw new UnsupportedAudioFileException(
                "This format is unsupported."
            );
        }

        final int channels = this.format.getChannels();
        final int nBits = this.format.getSampleSizeInBits();
        final boolean isBigEndian = this.format.isBigEndian();
        // Separate by channels
        final byte[] notSepByteArray = ais.readAllBytes();
        final byte[][] sepByteArray = separateByChannels(notSepByteArray);
        // byte -> double
        final double[][] sig = new double[channels][];
        for (int index = 0; index < channels; index++) {
            sig[index] =
                Converter.byte2double(sepByteArray[index], nBits, isBigEndian);
        }
        this.signal = adjustLength(sig);
    }


    /**
     * Output the member variable (double[][] signal) as a WAV file.
     *
     * @param filename 
     *      the name of the WAV file to be created
     * @throws IOException 
     *      if an I/O error occurs while writing the file
    */
    public void outputData(final String filename) throws IOException {
        // byte -> double
        final int channels = this.format.getChannels();
        final int nBits = this.format.getSampleSizeInBits();
        final boolean isBigEndian = this.format.isBigEndian();
        final byte[][] notConBytes = new byte[channels][];
        for (int i = 0; i < channels; i++) {
            notConBytes[i] = Converter.double2byte(this.signal[i], nBits, isBigEndian);
        }

        // connect byte array
        int connectPos = 0;
        final int sampleSize = nBits / 8;
        final int doubleArrayLength = this.signal[0].length;
        final byte[] conBytes = new byte[channels * doubleArrayLength * sampleSize];
        for (int i = 0; i < doubleArrayLength * sampleSize; i += sampleSize) {
            for (int j = 0; j < channels; j++) {
                for (int k = 0; k < sampleSize; k++) {
                    conBytes[connectPos] = notConBytes[j][i + k];
                    connectPos++;
                }
            }
        }

        // output
        final var is = new ByteArrayInputStream(conBytes); 
        final var ais = new AudioInputStream(is, this.format, conBytes.length);
        AudioSystem.write(ais, AudioFileFormat.Type.WAVE, new File(filename));
    }


    /** print audio format to System.out (debug code). */
    public void printAudioFormat() {
        System.out.println("Encoding: " + this.format.getEncoding());
        System.out.println("isBigEndian: " + this.format.isBigEndian());        
        System.out.println("Channels:" + this.format.getChannels());
        System.out.println("");
        System.out.println("SampleRate: " + this.format.getSampleRate());
        System.out.println("BitDepth: " + this.format.getSampleSizeInBits());
        System.out.println("");
        System.out.println("FrameRate: " + this.format.getFrameRate());
        System.out.println("FrameSize: " + this.format.getFrameSize());
    }


    // private methods -----------------------------------------------------------
    private static boolean isFormatOK(AudioFormat f) {
        final int channels = f.getChannels();
        final int nBits = f.getSampleSizeInBits();

        if (
            f.getEncoding() == AudioFormat.Encoding.PCM_SIGNED
                && (channels == 1 || channels == 2)
                && (nBits == 8 || nBits == 16 || nBits == 24 || nBits == 32)
        ) {
            return true;
        } else { 
            return false;
        }
    }


    private static double[][] adjustLength(final double[][] a) {

        final List<Integer> length = new ArrayList<>();
        for (final double[] array : a) {
            length.add(array.length);
        }
        final int maxLength = Collections.max(length);

        final double[][] ret = new double[a.length][maxLength];
        for (int i = 0; i < ret.length; i++) {
            System.arraycopy(a[i], 0, ret[i], 0, a[i].length);
        }
    
        return ret;
    }


    private byte[][] separateByChannels(final byte[] byteArray) {
        final int channels = this.format.getChannels();
        final int frameSize = this.format.getFrameSize();
        final int sampleSize = frameSize / channels;
        final int[] pos = new int[channels];
        final byte[][] sepByteArray = new byte[channels][byteArray.length / channels];

        if (channels == 2) {
            for (int i = 0; i < byteArray.length; i += frameSize) {
                for (int j = 0; j < sampleSize; j++) {
                    sepByteArray[0][pos[0]] = byteArray[i + j];
                    pos[0]++;
                }   
                for (int j = sampleSize; j < frameSize; j++) {
                    sepByteArray[1][pos[1]] = byteArray[i + j];
                    pos[1]++;
                }   
            }
        } else { // channels == 1
            sepByteArray[0] = byteArray;
        }

        return sepByteArray;
    }
}
